### DISCUSSION TITLE
Hashtag mysteriously appears in post

### DISCUSSION THREAD

**Discussion by FediScienceSupport** [mastodon/mastodon#32218] [Q&A] [answered]
Need some help with a baffling problem.

Someone wrote to let us know that their post was tagged with a tag that they did not add themselves.
The hashtag appeared beneath an image they had in their post.

Upon checking, it was clear that the tag was indeed not mentioned in the post's text, but it was assigned to it in statuses_tags.

I deleted the entry from statuses_tags and it disappeared from the post. But when they edited the status it reappeared. 
They also tried to use a different client on a different device and the same thing happened.

This happened the first time after some maintenance was done on the Mastodon server during which the docker containers where stopped and started multiple times.

Any idea what server side process could be doing this?

**Comment by renchap** [comment]
This is very weird. Could you share the status?

Something that often happens is statuses created from other software than Mastodon having what we call "out-of-band" hashtags, which are not in the status text, but added as metadata. Mastodon started displaying those in a small bar at the bottom of the posts in version 4.2.

**Reply by FediScienceSupport** [reply]
The poster has used the vanilla web interface to make the posts. But yeah I have seen the out-of-band tags happening with e.g. pixelfed posts when shown on a Mastodon site.

The tag injection just happened again here: https://fediscience.org/@fairsharing/113276952015114501 but I already removed the post/tag relationship and the tag no longer shows.

At this point I am just hoping this will go away when we update to 4.3.

**Reply by ClearlyClaire** [reply]
Let us know if it happens again, without altering the database since it prevents us from understanding what is happening.

**Reply by FediScienceSupport** [reply]
Thank you. We updated to 4.3 and the problem persists.
See https://fediscience.org/@fairsharing/113232915863390653
This is an old post from which I removed the tag previously. I asked the poster to edit it after we updated to 4.3 and with that edit the tag appeared again.
Since the tag is not in the content body (never has been), where does the system take it from? There must be some record somewhere that relates the tag to the post, right?

**Reply by ClearlyClaire** [reply]
There seem to be something wrong in your database with `#DatabasesDemystified` and `#LiberalDemocraticParty` being conflated into one, though I'm not sure how this could happen.

Can you open a Rails console (`RAILS_ENV=production bundle exec rails c`) and check the following?
```ruby
Status.find(113232915863390653).tags.pluck(:id, :name)
Tag.find_normalized('liberaldemocraticparty')
```

**Reply by FediScienceSupport** [reply]
I think I see the problem now. find_normalized for 'databasesdemystified' returns the 'liberaldemocraticparty' tag.

does this function use some kind of search index that I can rebuild?

```
irb(main):001> Status.find(113232915863390653).tags.pluck(:id, :name)

=> [[522667, "liberaldemocraticparty"], [24739, "databases"], [159338, "datacuration"], [974930, "fairsharingcommunitychampions"]]

irb(main):002> Tag.find_normalized('liberaldemocraticparty')
irb(main):003>
#<Tag:0x00007fc7231bb5b0
 id: 522667,
 name: "liberaldemocraticparty",
 created_at: Sat, 10 Dec 2022 07:30:08.370374000 UTC +00:00,
 updated_at: Sat, 05 Oct 2024 07:19:29.056540000 UTC +00:00,
 usable: nil,
 trendable: nil,
 listable: nil,
 reviewed_at: nil,
 requested_review_at: nil,
 last_status_at: Sat, 05 Oct 2024 07:19:27.000000000 UTC +00:00,
 max_score: nil,
 max_score_at: nil,
 display_name: "liberaldemocraticparty">

irb(main):004> Tag.find_normalized('databasesdemystified')
#<Tag:0x00007fc7221f1af8
 id: 522667,
 name: "liberaldemocraticparty",
 created_at: Sat, 10 Dec 2022 07:30:08.370374000 UTC +00:00,
 updated_at: Sat, 05 Oct 2024 07:19:29.056540000 UTC +00:00,
 usable: nil,
 trendable: nil,
 listable: nil,
 reviewed_at: nil,
 requested_review_at: nil,
 last_status_at: Sat, 05 Oct 2024 07:19:27.000000000 UTC +00:00,
 max_score: nil,
 max_score_at: nil,
 display_name: "liberaldemocraticparty">
```

**Reply by ClearlyClaire** [reply]
It seems like at least one of your database indexes is corrupted, which is bad news.

Which version of PostgreSQL are you using?

Can you try the following in SQL?
```sql
CREATE EXTENSION IF NOT EXISTS amcheck;
SELECT bt_index_check(c.oid)
FROM pg_index i
JOIN pg_class c ON i.indexrelid = c.oid
WHERE c.relname = 'index_tags_on_name_lower_btree';
```

**Reply by FediScienceSupport** [reply]
14.10

Query returns
```
 bt_index_check
----------------

(1 row)
```

**Reply by ClearlyClaire** [reply]
Ok, what about `SELECT bt_index_parent_check(c.oid) FROM pg_index i JOIN pg_class c ON i.indexrelid = c.oid WHERE c.relname = 'index_tags_on_name_lower_btree';`?

**Reply by FediScienceSupport** [reply]
```
 bt_index_parent_check
-----------------------

(1 row)
```

**Reply by ClearlyClaire** [reply, marked as answer]
Thanks, can you also check:
```sql
SELECT bt_index_parent_check(c.oid, true) FROM pg_index i JOIN pg_class c ON i.indexrelid = c.oid WHERE c.relname = 'index_tags_on_name_lower_btree';
```

Regardless of what it returns (though I'm still interested in that), can you try `REINDEX INDEX CONCURRENTLY index_tags_on_name_lower_btree;`?

**Reply by FediScienceSupport** [reply]
Thank **you**.

```
 bt_index_parent_check
-----------------------

(1 row)
```


Done the reindex.

`Tag.find_normalized('databasesdemystified')`

now returns nil

**Reply by ClearlyClaire** [reply]
Thanks. The fact that `amcheck` is not catching that is concerning :/

But it does sound like the PostgreSQL 14.0 to 14.3 index corruption issue.

**Comment by FediScienceSupport** [comment]
If I unlink the 'liberaldemocraticparty' tag from the post and  I ask the poster to edit the post again, the system should then create the 'databasesdemystified' tag and link it properly. Or is there a better way to now get this to the correct state?

**Reply by ClearlyClaire** [reply]
Yes, that should work.

**Reply by FediScienceSupport** [reply]
Many thanks for your help. It's really appreciated.

**Reply by ClearlyClaire** [reply]
You might want to `REINDEX` the whole of `mastodon_production` since it is likely other indexes are corrupted as well.

**Reply by FediScienceSupport** [reply]
Yeah. I think I will do that. Thank you.
